lib/logic/player/favorites.dart

```diff
@@ -1,27 +1,133 @@
+import 'package:firebase_crashlytics/firebase_crashlytics.dart';
+import 'package:flutter/foundation.dart';
 import 'package:sweyer/sweyer.dart';
 
-// class FavoriteRepository {
-//   FavoriteRepository._();
-//   static final instance = FavoriteRepository._();
+@visibleForTesting
+class FavoritesRepository {
+  final serializersMap = ContentMap<IntSerializerType>({
+    for (final contentType in Content.enumerate())
+      contentType: IntListSerializer('favorites_${ContentUtils.contentTypeString(contentType)}.json')
+  });
+}
 
-//   final _serializers = ContentMap<IntListSerializer>();
+class FavoritesControl extends Control {
+  static FavoritesControl instance = FavoritesControl();
 
-//   @override
-//   Future<void> markFavorite<T extends Content>(bool value) async {
-//     assert(T != Content);
-//     _serializers.getValue<T>();
-//   }
-// }
+  @visibleForTesting
+  final repository = FavoritesRepository();
 
-// class _FavoriteBatch {
-//   bool committed = false;
+  final _favoriteSetsMap = ContentMap<Set<int>>();
 
-//   void commit() {
-//     assert(!committed);
-//     committed = true;
-//   }
+  bool _useAndroidFavorites(Type contentType) =>
+    contentType == Song && DeviceInfoControl.instance.useScopedStorageForFileModifications;
```

Comment:
I don't feel like we should have this option, since it leaves a question on what we should do when this option is being switch on and off, it's also a bit too specific, since it causes the loading dialog only on Android 30. It's better to just leave it on.

Yes maybe on some API versions we will have this loading dialog, but that is how the system handles it and we cannot really do anything with it, this is not up to us. To be honest, it seems to me not too bad, I expected worse - a similar confirmation dialog to what is shown when you try to delete the songs and was really relieved when it was just a small loading.

Reply:
> it's also a bit too specific, since it causes the loading dialog only on Android 30.

This is from my tests on an emulator, there might very well be other devices out there.

> To be honest, it seems to me not too bad,

I strongly disagree! This dialog is extremely disruptive and annoying, given the fact that it doesn't really show anything and just loads for a second. This is not at all like the deletion dialog, where a confirmation definitely makes sense. For this, not at all. I myself (and probably many others) don't really care if the file is marked as favorite in the MediaStore, the important part is that it is marked in Sweyer. Please add the option to disable this!

>  but that is how the system handles it and we cannot really do anything with it, this is not up to us.

Exactly, the only option we have is to allow the user to disable the integration, so we can use the favorites in the app without interruptions.

>  a question on what we should do when this option is being switch on and off

For now, if it is turned off, do nothing, if it is turned on just update the favorites just like we do on startup. This can be improved in the future, but the option to disable it is more important to have.

Reply:
I just realized that we don't use the local `FavoritesRepository` when we use the `MediaStore`, so when the option is disabled then we should just store the current favorites in the local `FavoritesRepository`.

Reply:
> For now, if it is turned off, do nothing, if it is turned on just update the favorites just like we do on startup. This can be improved in the future, but the option to disable it is more important to have.

So there will be two independent sources which will not synchronize with each other when the option is switched on and off - is that what you are saying?

Reply:
When switching it off, you can just trigger a save to the `FavoritesRepository`, so all current loaded favorites from the media store are synchronized. When turning it on, it would be nice to update all favorites from the `FavoritesRepository` in the media store, but for now I think it is good enough to just ignore the `FavoritesRepository` and load from the media store just like we do on startup.

Reply:
> When switching it off, you can just trigger a save to the FavoritesRepository, so all current loaded favorites from the media store are synchronized

Just triggering a save will lose any already marked favorites, that's what I'm definitely not going to do. The same applies if we are switching from local source to Android source - just making it the same as our current local source will lose any data in the Android's source.

I'm just going to leave two independent sources, there's no way to safely merge those sources, since there's no source of truth if we create such an option.

Reply:
> Just triggering a save will lose any already marked favorites,

Correct me if I'm wrong, but aren't the favorites stored in the `_favoriteSetsMap` when we load them at startup from the media store? If we just save this map to the repository when turning off the integration, no favorites would be lost.

This doesn't work when enabling the integration, we would loose all local favorites. As I mentioned that could be improved in the future, letting the user choose the source of truth.

If you think this is too complicated for now, just add the option and leave them independent, we can always improve this later.